File: 47. Permutations II.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums = sorted(nums)
        self.res = []
        used = [False for i in xrange(len(nums))]
        self.DFS([], used, nums)
        return self.res
        
    def DFS(self, path, used, nums):
        if len(path) == len(nums):
            self.res+=[path[:]]
            return
        for i in xrange(len(nums)):
            if not used[i]:
                if i>0 and not used[i-1] and nums[i] == nums[i-1]:
                    continue
                path += [nums[i]]
                used[i] = True
                self.DFS(path, used, nums)
                path.pop()
                used[i] = False

File: 47. Permutations II.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> path;
        vector<int> used (nums.size(), false);
        DFS(res, path, nums, used);
        return res;
    }
    void DFS(vector<vector<int>> &res, vector<int> &path, vector<int> &nums, vector<int> &used)
    {
        if (path.size() == nums.size())
        {
            res.push_back(path);
            return;
        }
        for (int i=0; i<nums.size(); i++)
        {
            if (!used[i])
            {
                if (i>0 && !used[i-1] && nums[i] == nums[i-1])
                    continue;
                used[i] = true;
                path.push_back(nums[i]);
                DFS(res, path, nums, used);
                path.pop_back();
                used[i] = false;
            }
        }
    }
};
